WcaOnRails/app/webpacker/javascript/search_widget/OmnisearchInput.js

```diff
@@ -0,0 +1,149 @@
+import React, {
+  useState, useEffect, useCallback,
+} from 'react';
+import { Dropdown } from 'semantic-ui-react';
+
+import CompetitionItem from './CompetitionItem';
+import RegulationItem from './RegulationItem';
+import UserItem from './UserItem';
+import TextItem from './TextItem';
+import useDebounce from '../hooks/useDebounce';
+import I18n from '../i18n';
+import { fetchJsonOrError } from '../requests/fetchWithAuthenticityToken';
+import './OmnisearchInput.scss';
+
+const classToComponent = {
+  user: UserItem,
+  person: UserItem,
+  competition: CompetitionItem,
+  regulation: RegulationItem,
+  text: TextItem,
+};
+
+const ItemFor = ({ item }) => {
+  const Component = classToComponent[item.class];
+  return (
+    <div className="selected-item">
+      <Component item={item} />
+    </div>
+  );
+};
+
+const renderLabel = ({ item }) => ({
+  color: 'blue',
+  content: <ItemFor item={item} />,
+  className: 'omnisearch-item',
+  as: 'div',
+});
+
+const itemToOption = (item) => ({
+  item,
+  id: item.id,
+  key: item.id,
+  value: item.id,
+  // 'text' is used by the search method from the component, we need to put
+  // the text with a potential match here!
+  text: [item.id, item.name, item.content_html, item.search].join(' '),
+  content: <ItemFor item={item} />,
+});
+
+const createSearchItem = (search) => itemToOption({
+  class: 'text',
+  id: 'search',
+  url: `/search?q=${encodeURIComponent(search)}`,
+  search,
+});
+
+const DEBOUNCE_MS = 300;
+
+const OmnisearchInput = ({
+  url,
+  goToItemOnSelect,
+  placeholder,
+  removeNoResultsMessage,
+}) => {
+  const [search, setSearch] = useState('');
+  const [results, setResults] = useState([]);
+  const [loading, setLoading] = useState(false);
+  const debouncedSearch = useDebounce(search, DEBOUNCE_MS);
+
+  const [selected, setSelected] = useState([]);
+
+  const handleChange = useCallback((e, { value, options }) => {
+    setSearch('');
+    // Here we have "value" which contains the ids of the elements selected,
+    // "oldSelected" which contains the previously selected elements,
+    // "options" which contains the currently displayed options.
+    // "options" changes over time, and may not contain previously selected
+    // elements anymore: we need to make sure the new "selected" value includes
+    // all elements details for the elements in "value", they may come either
+    // from "oldSelected" or "options".
+    setSelected((oldSelected) => oldSelected.concat(options).filter(({ id }) => {
+      const index = value.indexOf(id);
+      if (index !== -1) {
+        value.splice(index, 1);
```

Comment:
This comes from the fact that we may have duplicates in `oldSelected.concat(options)`, right? In that case I think it would be slightly cleaner to do:
```js
const unique = (xs) => [...new Set(xs)];
// ...
unique(oldSelected.concat(options)).filter(({ id }) => value.includes(id));
```

Reply:
Precisely!
Thanks I totally forgot `Set` was a thing in javascript, it's indeed much cleaner this way.